plugins/GSdx/Renderers/HW/GSTextureCache.cpp

```diff
@@ -2108,6 +2058,194 @@ void GSTextureCache::AttachPaletteToSource(Source* s, uint16 pal, bool need_gs_t
+
+	// Key parameter is valid.
+	auto it = m_surface_offset_cache.find(sok);
+	if (it != m_surface_offset_cache.end())
+		return it->second;  // Cache HIT.
+	
+	// Cache MISS.
+	// Search for a valid <x,y> offset from B to A in B coordinates.
+	SurfaceOffset so;
+	so.is_valid = false;
+	const int dx = b_psm_s.bs.x;
+	const int dy = b_psm_s.bs.y;
+	GSVector4i b2a_offset = GSVector4i::zero();
+	if (a_el.bp >= b_el.bp)
+	{
+		// A starts after B, search <x,y> offset from B to A in B coords.
+		for (b2a_offset.x = b_el.rect.x; b2a_offset.x < b_el.rect.z; b2a_offset.x += dx)
+		{
+			for (b2a_offset.y = b_el.rect.y; b2a_offset.y < b_el.rect.w; b2a_offset.y += dy)
+			{
+				const uint32 a_candidate_bp = b_psm_s.bn(b2a_offset.x, b2a_offset.y, b_el.bp, b_el.bw);
+				if (a_el.bp == a_candidate_bp)
+				{
+					so.is_valid = true;  // Sweep search HIT: <x,y> offset found.
+					break;
+				}
+			}
+			if (so.is_valid)
+				break;
+		}
+	}
+	else
+	{
+		// B starts after A, suppose <x,y> offset is <x,y> of the B validity rectangle.
+		so.is_valid = true;
+		b2a_offset.x = b_el.rect.x;
+		b2a_offset.y = b_el.rect.y;
+	}
+
+	assert(!so.is_valid || b2a_offset.x >= b_el.rect.x);
+	assert(!so.is_valid || b2a_offset.x < b_el.rect.z);
+	assert(!so.is_valid || b2a_offset.y >= b_el.rect.y);
+	assert(!so.is_valid || b2a_offset.y < b_el.rect.w);
+
+	if (so.is_valid)
+	{
+		// Search for a valid <z,w> offset from B to the end of A in B coordinates.
+		if (a_bp_end >= b_bp_end)
+		{
+			// A ends after B, suppose <z,w> offset is <z,w> of the B validity rectangle.
+			b2a_offset.z = b_el.rect.z;
+			b2a_offset.w = b_el.rect.w;
+		}
+		else
+		{
+			// B ends after A, sweep search <z,w> offset in B coordinates.
+			so.is_valid = false;
+			for (b2a_offset.z = b2a_offset.x + dx - 1; b2a_offset.z <= b_el.rect.z - 1 + dx; b2a_offset.z += dx)
+			{
+				for (b2a_offset.w = b2a_offset.y + dy - 1; b2a_offset.w <= b_el.rect.w - 1 + dy; b2a_offset.w += dy)
+				{
+					const uint32 a_candidate_bp_end = b_psm_s.bn(b2a_offset.z, b2a_offset.w, b_el.bp, b_el.bw);
+					if (a_bp_end == a_candidate_bp_end)
+					{
+						// Align b2a_offset outside.
+						if (b2a_offset.z < b_el.rect.z)
+							++b2a_offset.z;
+						else
+							b2a_offset.z = b_el.rect.z;
+						if (b2a_offset.w < b_el.rect.w)
+							++b2a_offset.w;
+						else
+							b2a_offset.w = b_el.rect.w;
+						so.is_valid = true;  // Sweep search HIT: <z,w> offset found.
+						break;
+					}
+				}
+				if (so.is_valid)
+					break;
+			}
+			if (!so.is_valid)
+			{
+				// Sweep search <z,w> MISS.
+				GL_CACHE("TC: ComputeSurfaceOffset - Could not find <z,w> offset.");
+				b2a_offset.z = b_el.rect.z;
+				b2a_offset.w = b_el.rect.w;
+			}
+		}
+	}
+
+	assert(!so.is_valid || b2a_offset.z > b2a_offset.x);
+	assert(!so.is_valid || b2a_offset.z <= b_el.rect.z);
+	assert(!so.is_valid || b2a_offset.w > b_el.rect.y);
+	assert(!so.is_valid || b2a_offset.w <= b_el.rect.w);
+	assert(!so.is_valid || b2a_offset.z % dx == 0);
```

Comment:
Reminder to myself: this is not necessarily true, remove assertion here and below.